python/lsst/afw/image/readers.cc

```diff
@@ -0,0 +1,310 @@
+}
+
+void declareImageFitsReader(py::module & mod) {
+    PyImageFitsReader cls(mod, "ImageFitsReader");
+    declareCommonMethods(cls);
+    declareSinglePlaneMethods(cls);
+    cls.def(
+        "read",
+        [](ImageFitsReader & self, lsst::geom::Box2I const & bbox, ImageOrigin origin, py::object dtype) {
+            if (dtype == py::none()) {
+                dtype = py::dtype(self.readDType());
+            }
+            return utils::python::TemplateInvoker::apply(
+                makeReadHelper([&](auto t) { return self.read<decltype(t)>(bbox, origin); }),
+                py::dtype(dtype),
+                utils::python::TemplateInvoker::Tag<std::uint16_t, int, float, double, std::uint64_t>()
+            );
+        },
+        "bbox"_a=lsst::geom::Box2I(), "origin"_a=PARENT, "dtype"_a=py::none()
+    );
+}
+
+void declareMaskFitsReader(py::module & mod) {
+    PyMaskFitsReader cls(mod, "MaskFitsReader");
+    declareCommonMethods(cls);
+    declareSinglePlaneMethods(cls);
+    cls.def(
+        "read",
+        [](MaskFitsReader & self, lsst::geom::Box2I const & bbox, ImageOrigin origin,
+           bool conformMasks, py::object dtype) {
+            if (dtype == py::none()) {
+                dtype = py::dtype(self.readDType());
+            }
+            return utils::python::TemplateInvoker::apply(
+                makeReadHelper([&](auto t) { return self.read<decltype(t)>(bbox, origin, conformMasks); }),
+                py::dtype(dtype),
+                utils::python::TemplateInvoker::Tag<MaskPixel>()
+            );
+        },
+        "bbox"_a=lsst::geom::Box2I(), "origin"_a=PARENT, "conformMasks"_a=false, "dtype"_a=py::none()
+    );
+    // all other methods provided by base class wrappers
+}
+
+void declareMaskedImageFitsReader(py::module & mod) {
+    PyMaskedImageFitsReader cls(mod, "MaskedImageFitsReader");
+    cls.def(py::init<std::string const &, int>(), "fileName"_a, "hdu"_a=fits::DEFAULT_HDU);
+    cls.def(py::init<fits::MemFileManager&, int>(), "manager"_a, "hdu"_a=fits::DEFAULT_HDU);
+    declareCommonMethods(cls);
+    declareMultiPlaneMethods(cls);
+    cls.def("readPrimaryMetadata", &MaskedImageFitsReader::readPrimaryMetadata);
+    cls.def("readImageMetadata", &MaskedImageFitsReader::readImageMetadata);
+    cls.def("readMaskMetadata", &MaskedImageFitsReader::readMaskMetadata);
+    cls.def("readVarianceMetadata", &MaskedImageFitsReader::readVarianceMetadata);
+    cls.def(
+        "read",
+        [](MaskedImageFitsReader & self, lsst::geom::Box2I const & bbox, ImageOrigin origin,
+           bool conformMasks, bool needAllHdus, py::object dtype) {
+            if (dtype == py::none()) {
+                dtype = py::dtype(self.readImageDType());
+            }
+            return utils::python::TemplateInvoker::apply(
+                makeReadHelper([&](auto t) {
+                    return self.read<decltype(t)>(bbox, origin, conformMasks);
+                }),
+                py::dtype(dtype),
```

Comment:
It shouldn't be a big deal in this case, because the pybind11 constructor should be able to see that the Python type is already correct.  I think any other approach it this would require another local variable (in every instance of this pattern), which I'd like to avoid.